// @vuex-orm/core/lib/modules/RootMutations.js
import Query from '../query/Query';
import OptionsBuilder from './support/OptionsBuilder';
/**
 * Delete records from the store. The actual name for this mutation is
 * `delete`, but named `destroy` here because `delete` can't be declared at
 * this scope level.
 */
function destroy(state, payload) {
    var entity = payload.entity;
    var where = payload.where;
    var result = payload.result;
    result.data = (new Query(state, entity)).delete(where);
}
/**
 * Delete all data from the store.
 */
function deleteAll(state, payload) {
    if (payload && payload.entity) {
        (new Query(state, payload.entity)).deleteAll();
        return;
    }
    Query.deleteAll(state);
}
var RootMutations = {
    /**
     * Execute generic mutation. This method is used by `Model.commit` method so
     * that user can commit any state changes easily through models.
     */
    $mutate: function (state, payload) {
        payload.callback(state[payload.entity]);
    },
    /**
     * Create new data with all fields filled by default values.
     */
    new: function (state, payload) {
        var entity = payload.entity;
        var result = payload.result;
        result.data = (new Query(state, entity)).new();
    },
    /**
     * Save given data to the store by replacing all existing records in the
     * store. If you want to save data without replacing existing records,
     * use the `insert` method instead.
     */
    create: function (state, payload) {
        var entity = payload.entity;
        var data = payload.data;
        var options = OptionsBuilder.createPersistOptions(payload);
        var result = payload.result;
        result.data = (new Query(state, entity)).create(data, options);
    },
    /**
     * Insert given data to the state. Unlike `create`, this method will not
     * remove existing data within the state, but it will update the data
     * with the same primary key.
     */
    insert: function (state, payload) {
        var entity = payload.entity;
        var data = payload.data;
        var options = OptionsBuilder.createPersistOptions(payload);
        var result = payload.result;
        result.data = (new Query(state, entity)).insert(data, options);
    },
    /**
     * Update data in the store.
     */
    update: function (state, payload) {
        var entity = payload.entity;
        var data = payload.data;
        var where = payload.where || null;
        var options = OptionsBuilder.createPersistOptions(payload);
        var result = payload.result;
        result.data = (new Query(state, entity)).update(data, where, options);
    },
    /**
     * Insert or update given data to the state. Unlike `insert`, this method
     * will not replace existing data within the state, but it will update only
     * the submitted data with the same primary key.
     */
    insertOrUpdate: function (state, payload) {
        var entity = payload.entity;
        var data = payload.data;
        var options = OptionsBuilder.createPersistOptions(payload);
        var result = payload.result;
        result.data = (new Query(state, entity)).insertOrUpdate(data, options);
    },
    delete: destroy,
    deleteAll: deleteAll
};
export default RootMutations;
